src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1221,650 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
 			NULL /* pgResult */);
 	}
 }
+
+
+/*
+ * SplitShard API to split a given shard (or shard group) in non-blocking fashion
+ * based on specified split points to a set of destination nodes.
+ * 'splitOperation'             : Customer operation that triggered split.
+ * 'shardIntervalToSplit'       : Source shard interval to be split.
+ * 'shardSplitPointsList'		: Split Points list for the source 'shardInterval'.
+ * 'workersForPlacementList'	: Placement list corresponding to split children.
+ */
+static void
+NonBlockingShardSplit(SplitOperation splitOperation,
+					  ShardInterval *shardIntervalToSplit,
+					  List *shardSplitPointsList,
+					  List *workersForPlacementList)
+{
+	char *superUser = CitusExtensionOwnerName();
+	char *databaseName = get_database_name(MyDatabaseId);
+	int connectionFlags = FORCE_NEW_CONNECTION;
+
+	List *sourceColocatedShardIntervalList = ColocatedShardIntervalList(
+		shardIntervalToSplit);
+
+	/* First create shard interval metadata for split children */
+	List *shardGroupSplitIntervalListList = CreateSplitIntervalsForShardGroup(
+		sourceColocatedShardIntervalList,
+		shardSplitPointsList);
+
+	/* Only single placement allowed (already validated RelationReplicationFactor = 1) */
+	List *sourcePlacementList = ActiveShardPlacementList(shardIntervalToSplit->shardId);
+	Assert(sourcePlacementList->length == 1);
+	ShardPlacement *sourceShardPlacement = (ShardPlacement *) linitial(
+		sourcePlacementList);
+	WorkerNode *sourceShardToCopyNode = FindNodeWithNodeId(sourceShardPlacement->nodeId,
```

Comment:
would be nice to introduce a utility function to get the WorkerNode given a shard ID, because these same lines appear elesewhere

Reply:
@marcocitus : Introduced ActiveShardPlacementWorkerNode that you have in the draft PR for non blocking create_distributed_table.